### PR TITLE
fix: KyberDMM pick best pools, Polygon KyberDMM

### PR DESCRIPTION
**Pull request by romainbutteaud**
## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Review by dekz — APPROVED** [review]

File: packages/asset-swapper/contracts/src/KyberDmmSampler.sol [hunk 1]
```diff
@@ -20,12 +20,25 @@
...
 interface IKyberDmmFactory {
 
-    function getPoolAtIndex(address token0, address token1, uint256 index)
+    function getPools(address token0, address token1)
         external
         view
-        returns (address);
+        returns (address[] memory _tokenPools);
+
+    function getPoolsLength(address token0, address token1)
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/contracts/src/KyberDmmSampler.sol:38]: Used?

File: packages/asset-swapper/contracts/src/KyberDmmSampler.sol [hunk 2]
```diff
@@ -140,17 +153,25 @@ contract KyberDmmSampler
...
+                factory.getPools
                     {gas: KYBER_DMM_CALL_GAS}
-                    (path[i], path[i + 1], 0)
-                returns (address pool)
+                    (path[i], path[i + 1])
+                returns (address[] memory allPools)
             {
-                pools[i] = pool;
+                uint256 maxSupply = 0;
+                for (uint256 j = 0; j < allPools.length; j++) {
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/contracts/src/KyberDmmSampler.sol:168]: What happens if `allPools.length == 0`? We just have an invalid pool (null) at that index? Should we revert?

File: packages/asset-swapper/contracts/src/KyberDmmSampler.sol (hunk 1 above)
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/KyberDmmSampler.sol:38]: removing it

File: packages/asset-swapper/contracts/src/KyberDmmSampler.sol (hunk 2 above)
Comment (romainbutteaud) [reply packages/asset-swapper/contracts/src/KyberDmmSampler.sol:164]: fixed it, added a revert if no pools are found

**Pull request merged**